csrc/ops/arith.cpp

```diff
@@ -2528,47 +2528,58 @@ TopKResult topk(
       "TopKOp expects int64_t input for k but got ",
       k->dtype());
 
-  std::vector<IterDomain*> out_domain;
-  out_domain.reserve(inp_domain.size());
+  TensorView* out_values = nullptr;
+  TensorView* out_indices = nullptr;
+
+  // Create a root-to-logical resize for the topk dimension. The root
+  // dimension just inherits the same properties as the producer
+  // dimension. It's resized to generate a logical iter domain of
+  // extent K by slicing the root iter domain by [0:k].
+  //
+  // The first output generated when i == 0 is the value output. The
+  // second is the index output.
+  for (const int i : arange(2)) {
+    std::vector<IterDomain*> values_root;
+    values_root.reserve(inp_domain.size());
+    std::vector<IterDomain*> values_logical;
+    values_logical.reserve(inp_domain.size());
+
+    for (const auto [index, inp_domain_ptr] : enumerate(inp_domain)) {
+      auto root_id = inp_domain_ptr->cloneWithoutRFactor();
+      values_root.push_back(root_id);
+      if (index != (size_t)dim) {
+        // Root and logical are the same for non topk dim
+        values_logical.push_back(root_id);
+        continue;
+      }
 
-  for (const auto [index, inp_domain_ptr] : enumerate(inp_domain)) {
-    // TODO: nvfuser enumerate implementation is not correct, it should return
-    // signed ints instead.
-    if (index != (size_t)dim) {
-      out_domain.push_back(inp_domain_ptr->cloneWithoutRFactor());
-      continue;
-    }
+      auto logical_id = IterDomain::resize(
+          root_id,
+          inp->fusion()->zeroVal(DataType::Index),
+          SimplifyingIrBuilder::subExpr(k, root_id->extent()),
+          /*mark_as_rfactor=*/true);
 
-    // Handling top k dimension, since the output extent is k.
-    ExpressionEvaluator ee;
-    PolymorphicValue ext = ee.evaluate(k);
+      values_logical.push_back(logical_id);
+    }
 
-    IterType iter_type;
-    if (ext.hasValue()) {
-      iter_type =
-          ext.as<int64_t>() == 1 ? IterType::Broadcast : IterType::Iteration;
+    auto dtype = i == 0 ? inp->getDataType().value() : DataType::Int;
```

Comment:
```suggestion
    auto dtype = i == 0 ? inp->dtype() : DataType::Int;
```
Off-topic: is there a reason to keep `getDataType()` or should we just remove it?

Reply:
I don't remember why I used `getDataType()` instead of `dtype()`. I don't know if we really need `getDataType`.